frontend/ui/fts.lua

```diff
@@ -0,0 +1,259 @@
+local PREFERRED_MONOTONIC_CLOCKID = C.CLOCK_MONOTONIC
+-- Ditto for REALTIME (for :realtime_coarse only, :realtime uses gettimeofday ;)).
+local PREFERRED_REALTIME_CLOCKID = C.CLOCK_REALTIME
+-- CLOCK_BOOTTIME is only available on Linux 2.6.39+...
+local HAVE_BOOTTIME = false
+if ffi.os == "Linux" then
+    -- Unfortunately, it was only implemented in Linux 2.6.32, and we may run on older kernels than that...
+    -- So, just probe it to see if we can rely on it.
+    local probe_ts = ffi.new("struct timespec")
+    if C.clock_getres(C.CLOCK_MONOTONIC_COARSE, probe_ts) == 0 then
+        -- Now, it usually has a 1ms resolution on modern x86_64 systems,
+        -- but it only provides a 10ms resolution on all my armv7 devices :/.
+        if probe_ts.tv_sec == 0 and probe_ts.tv_nsec <= 1000000 then
+            PREFERRED_MONOTONIC_CLOCKID = C.CLOCK_MONOTONIC_COARSE
+        end
+    end
+    logger.dbg("fts: Preferred MONOTONIC clock source is", PREFERRED_MONOTONIC_CLOCKID == C.CLOCK_MONOTONIC_COARSE and "CLOCK_MONOTONIC_COARSE" or "CLOCK_MONOTONIC")
+    if C.clock_getres(C.CLOCK_REALTIME_COARSE, probe_ts) == 0 then
+        if probe_ts.tv_sec == 0 and probe_ts.tv_nsec <= 1000000 then
+            PREFERRED_REALTIME_CLOCKID = C.CLOCK_REALTIME_COARSE
+        end
+    end
+    logger.dbg("fts: Preferred REALTIME clock source is", PREFERRED_REALTIME_CLOCKID == C.CLOCK_REALTIME_COARSE and "CLOCK_REALTIME_COARSE" or "CLOCK_REALTIME")
+
+    if C.clock_getres(C.CLOCK_BOOTTIME, probe_ts) == 0 then
+        HAVE_BOOTTIME = true
+    end
+    logger.dbg("fts: BOOTTIME clock source is", HAVE_BOOTTIME and "supported" or "NOT supported")
+
+    probe_ts = nil --luacheck: ignore
+end
+
+--[[--
+Fixed point time. Maps to a POSIX struct timeval (<sys/time.h>).
+]]
+local fts = {}
+
+--[[--
+Returns an fts time based on the current wall clock time.
+(e.g., gettimeofday / clock_gettime(CLOCK_REALTIME).
+
+This is a simple wrapper around util.gettime() to get all the niceties of a fts.
+If you don't need sub-second precision, prefer os.time().
+Which means that, yes, this is a fancier POSIX Epoch ;).
+
+@usage
+    local fts = require("ui/fts")
+    local fts_start = fts.realtime()
+    -- Do some stuff.
+    -- You can add and substract fts times
+    local fts_duration = fts.realtime() - fts_start
+
+@treturn fts fixed point time
+]]
+function fts.realtime()
+    local sec, usec = util.gettime()
+    return sec * FTS_PRECISION + usec
+end
+
+--[[--
+Returns an fts time based on the current value from the system's MONOTONIC clock source.
+(e.g., clock_gettime(CLOCK_MONOTONIC).)
+
+POSIX guarantees that this clock source will *never* go backwards (but it *may* return the same value multiple times).
+On Linux, this will not account for time spent with the device in suspend (unlike CLOCK_BOOTTIME).
+
+@treturn Tfts fixed point time
+]]
+function fts.monotonic()
+    local timespec = ffi.new("struct timespec")
```

Comment:
Could possibly make sense to re-use a module-local ffi object everywhere instead of creating one every time, since we never pass it as-is, and always create new Lua numbers ;).